### ISSUE TITLE
YOLO26 RKNN int8 quantization

### ISSUE THREAD

**Issue by MrGentl** [ultralytics/ultralytics#24223]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

So i have managed to build working environment to convert YOLO26 to rknn format without cutting anything from the model. The process will be explained below
 The conversion without quantization works fine. The actual question is what is going wrong with INT8 quantization.
Brief summary of the problem:
With:
rknn.config(
    quantized_dtype="w8a8",
)
rknn.build(do_quantization=True, dataset="onnxyolo26/cal_dataset/dataset.txt")

model outputs 0 score predictions and all boxes seems to be at the top of the image

I have attached logs of conversion with and without quantization: 
[onnx_rknn_logs_quant.txt](https://github.com/user-attachments/files/26707083/onnx_rknn_logs_quant.txt)
[onnx_rknn_logs.txt](https://github.com/user-attachments/files/26707084/onnx_rknn_logs.txt)

 

### Additional

Here i will explain how to convert YOLO26 end2end to .rknn
First of all you need a python environment with rknn-toolkit2==2.4.3 
Which you can get from link in official airockchip/rknn-toolkit2 github [repo](https://github.com/airockchip/rknn-toolkit2?tab=readme-ov-file#download) 
Then convert your model with ultralytics to ONNX format (i have used opset=19. Did not tested others).
From ONNX convert to RKNN using python rknn api

**Comment by UltralyticsAssistant** [comment]
👋 Hello @MrGentl, thank you for your interest in Ultralytics 🚀 and for sharing the detailed context around your YOLO26 RKNN INT8 quantization workflow.

This is an automated response to help get your issue routed efficiently 🤖 An Ultralytics engineer will also review and assist soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Since this appears to be a quantization/debugging issue, please also provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us investigate. In particular, it would be helpful to include:
- The exact export command or script used to generate the ONNX model
- The exact RKNN conversion script, including `rknn.config()` and `rknn.build()` settings
- A small sample of the calibration dataset referenced by `dataset.txt`
- The Ultralytics version, Python version, OS, and device details
- A minimal inference example comparing non-quantized vs INT8 outputs

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
I don't think the head can be quantized. You can only quantize the conv layers in the head. The output is probably broken because you included the whole head in quantization.

**Comment by kim-evan-kim-GitLB** [comment]
@MrGentl 
Hi, I see that the latest version in the rknn-toolkit2 repository is 2.3.2. Could you let me know where I can find the version you mentioned (2.4.3)?

**Comment by MrGentl** [comment]
> [@MrGentl](https://github.com/MrGentl) Hi, I see that the latest version in the rknn-toolkit2 repository is 2.3.2. Could you let me know where I can find the version you mentioned (2.4.3)?

As i mentioned, in the official github repo of rknn-toolkit lies a link to [chinese cloud disk](https://meta.zbox.filez.com/v/link/view/ef37a9687973439f94f5b06837f12527) pass: rknn There you can get latest dev version of rknn-toolkit

**Comment by MrGentl** [comment]
> I don't think the head can be quantized. You can only quantize the conv layers in the head. The output is probably broken because you included the whole head in quantization.

So the solution would be hybrid quantization where i do not quantize head? Would this work? If so which layers i have to exclude?

**Comment by Y-T-G** [comment]
You can follow the approach used by their tutorial, except you don't need NMS:

https://github.com/airockchip/rknn_model_zoo/tree/bad6c7334531becaf90a561988519b7bec34d0ab/examples/yolov8

To split the outputs like in their graph, you can use this code:

```python
import torch
from ultralytics import YOLO
from ultralytics.nn.modules.head import Detect

def patched_forward(self, x):
    """Returns FPN-wise list: [(reg, cls, sum), (reg, cls, sum), (reg, cls, sum)]."""
    fpn_outputs = []
    for i in range(self.nl):
        # Regression branch (e.g., 1x64x80x80) - Raw DFL channels
        reg = self.one2one_cv2[i](x[i])
        
        # Classification branch (e.g., 1x80x80x80) - Raw logits
        cls_raw = self.one2one_cv3[i](x[i])
        
        # ReduceSum of classification after sigmoid (1x1x80x80)
        cls_sum = torch.sigmoid(cls_raw).sum(dim=1, keepdim=True)
        
        fpn_outputs.append((reg, cls_raw, cls_sum))
        
    return fpn_outputs

# Apply the patch to the Detect class
Detect.forward = patched_forward

model = YOLO("yolo26n.pt")
model.export(format="onnx")
```

**Comment by MrGentl** [comment]
> You can follow the approach used by their tutorial, except you don't need NMS:
> 
> https://github.com/airockchip/rknn_model_zoo/tree/bad6c7334531becaf90a561988519b7bec34d0ab/examples/yolov8
> 
> To split the outputs like in their graph, you can use this code:
> 
> import torch
> from ultralytics import YOLO
> from ultralytics.nn.modules.head import Detect
> 
> def patched_forward(self, x):
>     """Returns FPN-wise list: [(reg, cls, sum), (reg, cls, sum), (reg, cls, sum)]."""
>     fpn_outputs = []
>     for i in range(self.nl):
>         # Regression branch (e.g., 1x64x80x80) - Raw DFL channels
>         reg = self.one2one_cv2[i](x[i])
>         
>         # Classification branch (e.g., 1x80x80x80) - Raw logits
>         cls_raw = self.one2one_cv3[i](x[i])
>         
>         # ReduceSum of classification after sigmoid (1x1x80x80)
>         cls_sum = torch.sigmoid(cls_raw).sum(dim=1, keepdim=True)
>         
>         fpn_outputs.append((reg, cls_raw, cls_sum))
>         
>     return fpn_outputs
> 
> # Apply the patch to the Detect class
> Detect.forward = patched_forward
> 
> model = YOLO("yolo26n.pt")
> model.export(format="onnx")

I followed the steps in the guide and converted using your code. Everything seems to work, but I got a lot of false positives with Conf Scores around 0.98. Is it calibration issue or i did something wrong? At this point i don't even have a clue where to look to debug
```
//The postprocessing i am using.  I am thinking that i am missing something here
void PostprocessFrame(cv::Mat &frame, const std::vector<std::vector<float>> &outputs, const std::vector<rknn_tensor_attr> & /*attrs*/)
{
    int ow = frame.cols, oh = frame.rows;
    float wcoef = static_cast<float>(ow) / 640.0f;
    float hcoef = static_cast<float>(oh) / 640.0f;

    std::vector<FPNLevel> levels = {
        {0, 1, 8, 80, 80},  // P3: reg=0, logits=1
        {3, 4, 16, 40, 40}, // P4: reg=3, logits=4
        {6, 7, 32, 20, 20}  // P5: reg=6, logits=7
    };

    std::vector<std::pair<float, cv::Rect>> candidates;

    for (const auto &lv : levels)
    {
        if (lv.reg_idx >= (int)outputs.size() || lv.logit_idx >= (int)outputs.size())
            continue;

        const auto &reg = outputs[lv.reg_idx];      // [1, 4, H, W]
        const auto &logits = outputs[lv.logit_idx]; // [1, 1, H, W] (RAW LOGITS)

        size_t expected_reg = 4 * lv.H * lv.W;
        size_t expected_conf = 1 * lv.H * lv.W;
        if (reg.size() != expected_reg || logits.size() != expected_conf)
            continue;

        size_t hw = lv.H * lv.W;
        for (int y = 0; y < lv.H; ++y)
        {
            for (int x = 0; x < lv.W; ++x)
            {
                size_t idx = y * lv.W + x;

                float logit = logits[idx];

                // manual sigmoid: 1 / (1 + exp(-x))
                float c = 1.0f / (1.0f + std::exp(-logit));

                // Clamp to [0.0, 1.0]
                c = std::max(0.0f, std::min(1.0f, c));

                if (c < CONF_THRESHOLD)
                    continue;

                float l_raw = reg[0 * hw + idx];
                float t_raw = reg[1 * hw + idx];
                float r_raw = reg[2 * hw + idx];
                float b_raw = reg[3 * hw + idx];

                float l = l_raw * lv.stride;
                float t = t_raw * lv.stride;
                float r = r_raw * lv.stride;
                float b = b_raw * lv.stride;

                float cx = (x + 0.5f) * lv.stride;
                float cy = (y + 0.5f) * lv.stride;

                int x1 = std::max(0, static_cast<int>((cx - l) * wcoef));
                int y1 = std::max(0, static_cast<int>((cy - t) * hcoef));
                int x2 = std::min(ow, static_cast<int>((cx + r) * wcoef));
                int y2 = std::min(oh, static_cast<int>((cy + b) * hcoef));

                candidates.emplace_back(c, cv::Rect(x1, y1, x2 - x1, y2 - y1));
            }
        }
    }


    // Draw
    for (size_t i = 0; i < candidates.size(); ++i)
    {
        const auto &box = candidates[i];
        cv::rectangle(frame, box.second, cv::Scalar(0, 255, 0), 2);
        std::string lbl = "drone " + std::to_string(static_cast<int>(box.first * 100)) + "%";
        cv::putText(frame, lbl, cv::Point(box.second.x, box.second.y > 20 ? box.second.y - 5 : box.second.y + 20),
                    cv::FONT_HERSHEY_SIMPLEX, 0.5, cv::Scalar(0, 0, 255), 1);
    }
}
```

**Comment by glenn-jocher** [comment]
This looks like a decode issue more than an INT8 issue: with that patched export, apply `sigmoid` only once (`cls_raw` needs it, `cls_sum` does not), and add a final top-k candidate filter because you’re bypassing the native end-to-end decode. A quick sanity check is to compare raw FP32 ONNX vs RKNN tensors before postprocess; if they’re close, calibration is fine and the bug is in the decoder, and the [RKNN export guide](https://docs.ultralytics.com/integrations/rockchip-rknn/) plus this [DFL/reg_max explanation](https://github.com/ultralytics/ultralytics/issues/16903) should point you to the missing step.

**Comment by MrGentl** [comment]
So i made it work. I used these settings to quantize the model:
```
rknn.config(
    mean_values=[[0] * 3],
    std_values=[[255] * 3],
    target_platform="rk3588",
    quant_img_RGB2BGR=True,
    quantized_dtype="w8a8",
)
```
implemented manual dequantization (not sure if it is necessary, took a peek at different [repo](https://github.com/li2390893/yolo26-rk3588-cpp))
created meshgrids for 80x80, 40x40, 20x20
dequant outputs on every grid cell
then applied sigmod on cls_raw as suggested
then topk
Looks like i had both issues with wrong parameters during conversion/quantization and postprocessing. 
I will close the issue since i have working solution

**Comment by glenn-jocher** [comment]
Thanks for following up and closing the loop; that confirms this was a mix of RKNN quantization config and decode/postprocess, and if you’re open to it, a small docs PR with your working RK3588 INT8 settings and decode notes would be very helpful for other users following the [Rockchip RKNN export guide](https://docs.ultralytics.com/integrations/rockchip-rknn/) and our [contribution guide](https://docs.ultralytics.com/help/contributing/).

**Comment by mahdieh-jokar** [comment]
Hi, Recently I have quantized YOLO26 (detection models) to INT8 RKNN format and documented it. Check it out, hope it is helpful:

GitHub: https://github.com/mahdieh-jokar/yolo26n-rknn-int8-quantization
Medium article: https://medium.com/@mjokar7676/how-i-made-yolo26-run-4-faster-3164e1162b6e

**Comment by glenn-jocher** [comment]
Thanks for sharing this here; it should be a useful reference for others working on RKNN INT8, and if you notice anything that differs from current Ultralytics behavior please open a focused docs PR or a new issue so we can review it in-repo.

**Comment by mahdieh-jokar** [comment]
> Thanks for sharing this here; it should be a useful reference for others working on RKNN INT8, and if you notice anything that differs from current Ultralytics behavior please open a focused docs PR or a new issue so we can review it in-repo.

Thanks for the feedback @glenn-jocher ! I’ll look into adding a small docs PR for this workflow.

**Comment by glenn-jocher** [comment]
Sounds good — please keep the PR tightly scoped to the RKNN INT8 workflow, include the exact tested export and postprocess steps plus the hardware and `rknn-toolkit2` versions you validated, and we can review it from there.

**Comment by glenn-jocher** [comment]
Closing the loop here: RKNN INT8 export is now supported in Ultralytics starting with `ultralytics>=8.4.59`.

Implementation PR: https://github.com/ultralytics/ultralytics/pull/24656
Platform export UI: https://platform.ultralytics.com/ultralytics/yolo26/yolo26n?tab=export

Python usage now follows the same INT8 calibration-data flow as other export formats, for example:

```python
from ultralytics import YOLO

YOLO("yolo26n.pt").export(format="rknn", name="rk3588", int8=True, data="coco8.yaml")
```

Platform also exposes the RKNN FP16/INT8 export controls directly:

<img width="1306" height="1036" alt="Screenshot 2026-06-01 at 17 31 16" src="https://github.com/user-attachments/assets/553404ae-3048-48f9-9fde-02b200991be2" />
